kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,426 @@
+                engine.as_ref(),
+                &table_schema,
+                predicate.clone(),
+            )?;
+            scanner.into_scan_batches(engine.clone())
+        })
+        // Iterator<DeltaResult<Iterator<DeltaResult<TableChangesScanData>>>> to
+        // Iterator<DeltaResult<DeltaResult<TableChangesScanData>>>
+        .flatten_ok()
+        // Iterator<DeltaResult<DeltaResult<TableChangesScanData>>> to
+        // Iterator<DeltaResult<TableChangesScanData>>
+        .map(|x| x?);
+    Ok(result)
+}
+
+// Gets the expression for generating the engine data in [`TableChangesScanData`].
+//
+// TODO: This expression is temporary. In the future it will also select `cdc` and `remove` actions
+// fields.
+fn get_add_transform_expr() -> Expression {
+    Expression::Struct(vec![
+        column_expr!("add.path"),
+        column_expr!("add.size"),
+        column_expr!("add.modificationTime"),
+        column_expr!("add.stats"),
+        column_expr!("add.deletionVector"),
+        Expression::Struct(vec![column_expr!("add.partitionValues")]),
+    ])
+}
+
+/// Processes a single commit file from the log to generate an iterator of [`TableChangesScanData`].
+/// The scanner operates in two phases that _must_ be performed in the following order:
+/// 1. Prepare phase [`prepare_table_changes`]: This performs one iteration over every
+///    action in the commit. In this phase, we do the following:
+///     - Find the timestamp from a `CommitInfo` action if it exists. These are generated when
+///       In-commit timestamps is enabled.
+///     - Ensure that reading is supported on any protocol updates
+///     - Ensure that Change Data Feed is enabled for any metadata update. See  [`TableProperties`]
+///     - Ensure that any schema update is compatible with the provided `schema`. Currently, schema
+///       compatibility is checked through schema equality. This will be expanded in the future to
+///       allow limited schema evolution.
+///     - Determine if there exist any `cdc` actions.
+///     - Constructs the remove deletion vector map from paths belonging to `remove` actions to the
+///       action's corresponding [`DvInfo`]. This map will be filtered to only contain paths that
+///       exists in another `add` action _within the same commit_. This corresponds to `remove_dv`
+///       in [`TableChangesScanData`].
+/// 2. Scan file generation phase [`LogReplayScanner::into_scan_batches`]: This performs another
+///    iteration over every action in the commit, and generates [`TableChangesScanData`]. It does
+///    so by transforming the actions using [`get_add_transform_expr`], and generating selection
+///    vectors with the following rules:
+///     - If a `cdc` action was found in the prepare phase, only `cdc` actions are selected
+///     - Otherwise, select `add` and `remove` actions. Note that only `remove` actions that do not
+///       share a path with an `add` action are selected.
```

Comment:
Don't we also need to ignore non-dataChange actions? I don't see any code for that?

Reply:
ooohhh this is a good question! I had forgotten about this, thank you! also note we do have to be careful not to ignore _all_ non-dataChange actions because I think cdc actions (which we need) always (usually?) have `dataChange=false`.

Reply:
Right... I meant add and remove (forgot that cdc actions are also considered "file actions")

Reply:
Hmm correct me if I'm wrong, but this should be [covered already](https://github.com/delta-io/delta-kernel-rs/pull/540/files#diff-5cefdc42eff6c5705b4a1f8dfb7295ce5c4048bd832a1a2791b92dfeafd70f14R428) in the `FileActionSelectionVisitor`

Reply:
```
            if getters[0].get_str(i, "add.path")?.is_some() {
                self.selection_vector[i] = !self.has_cdc_action;
            } else if let Some(path) = getters[1].get_str(i, "remove.path")? {
                self.selection_vector[i] =
                    !self.has_cdc_action && !self.remove_dvs.contains_key(path)
            } else {
                self.selection_vector[i] = getters[2].get_str(i, "cdc.path")?.is_some()
            };
```
the final case sets to true if it is a cdc action. Otherwise all other actions are set to false.

Reply:
The missing bit would be something like this, I think?
```rust
            if getters[0].get_str(i, "add.path")?.is_some() {
                let isDataChange = getters[XXX].get("add.dataChange");
                self.selection_vector[i] = isDataChange && !self.has_cdc_action;
```
From the [spec for dataChange field](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#add-file-and-remove-file):
> When `false` the logical file must already be present in the table or the records in the added file must be contained in one or more remove actions in the same version

Reply:
Let's add a test for this as well :)

Reply:
test got added I believe, right @OussamaSaoudi-db? (I had the same comment in test file I think)